### DISCUSSION TITLE
REST & lifx API commands for Smart Switch (4 Gang)

### DISCUSSION THREAD

**Discussion by hewfish** [delfick/photons#102] [Q&A]
Hi,

I'm trying to get my head around the API both REST and lifx cli for the 4 gang switch. I can turn all relays off and on, but I can't seem to trigger individual relays. Has anyone had any experience with it or can point me in the right direction? Ultimately I'm trying to integrate it with HomeAssistant so that I can automate turning on my lights and then controlling the LED's I have on the relay segment (non-lifx). 

I have no issue discovering the switch, but I've tried getting my head around the documentation to see if I could query the available options for a device, and can't really get my head around that, if someone could even point me at how to query a device capabilities, I'm sure I could figure out the rest.

Thanks,
Matt

**Comment by delfick** [comment]
mildly related convo over here https://github.com/delfick/photons/issues/100

> if someone could even point me at how to query a device capabilities

Essentially the process of determining what a product can do is you get the hardware and firmware version of the device and then match it up to an external registry of what that can do https://github.com/lifx/products#determining-capabilities

Different libraries have different ways of doing that.

I'm not sure exactly how that's done in homeassistant. @Djelibeybi might have a better idea.

**Reply by hewfish** [reply]
Thanks, that might give me some direction. I'll also check out the related conversation on homeassistant forums, I think I remember seeing that node-red article, and it may have the detail I need as well.

**Reply by Djelibeybi** [reply]
I'm almost done, give me two minutes to finish typing.

**Comment by Djelibeybi** [comment]
There are two commands used to `query` switches and `set` the relay states. 

Relays are indxed from 0 to 3, so to get the status of a the first relay attached a switch with the serial `d073d5123456`, use the `query` command to send a `GetRPower` packet, specifying the relay in the `relay_index` field:

```json
{
    "command": "query",
    "args": {
        "matcher": {
            "serial": "d073d5123456"
        },
        "timeout": 10,
        "pkt_type": "GetRPower",
        "pkt_args": {
            "relay_index": 0
        }
    }
}
```

You'll get a reply that looks like this:

```json
{
    "results": {
        "d073d5123456": {
            "payload": {
                "level": 65535,
                "relay_index": 0
            },
            "pkt_name": "StateRPower",
            "pkt_type": 818
        }
    }
}
```

The `level` field of the `payload` contains the current power state, where (like bulbs), `0` is off and `65535` is on. To turn that relay off, send a `SetRPower` packet using the `set` command:

```json
{
    "command": "set",
    "args": {
        "matcher": {
            "serial": "d073d5123456"
        },
        "timeout": 10,
        "pkt_type": "SetRPower",
        "pkt_args": {
            "relay_index": 0,
            "level": 0
        }
    }
}
```

Ignore the response you get from the issuing the command as it usually shows you the previous state. Send a `GetRPower` request to get the current state. 

Note that neither the switch nor Photons provides any protection against turning off a relay that is powering a LIFX device, so keep that in mind too. :)

**Reply by hewfish** [reply]
Awesome, exactly what I needed. Thankyou.

**Reply by Djelibeybi** [reply]
You're welcome. :)

**Comment by Djelibeybi** [comment]
For the CLI, to get the current power level of a relay:

```shell
lifx lan:attr d073d5123456 GetRPower --silent -- '{"relay_index": 0}' | jq
{
  "level": 65535,
  "relay_index": 0
}
```

To turn that relay off:

```shell
lifx lan:attr d073d5123456 SetRPower --silent -- '{"relay_index": 0, "level": 0}' | jq
{
  "level": 65535,
  "relay_index": 0
}
```

Note the response still says `65535`, which is because it's showing the previous state. Sending another `GetRPower` will show the current state.

**Comment by hewfish** [comment]
Also very helpful. Thanks both for your quick response and easy answers. I ran a CURL against the REST and it was practically instant. Great work :)

Off:
curl -XPUT http://127.0.0.1:6100/v1/lifx/command -HContent-Type:application/json -d '{"command": "set","args": {"matcher": {"serial": "d073d5635a15"},"timeout": 10,"pkt_type": "SetRPower","pkt_args": {"relay_index": 0,"level": 0}}}'

On:
curl -XPUT http://127.0.0.1:6100/v1/lifx/command -HContent-Type:application/json -d '{"command": "set","args": {"matcher": {"serial": "d073d5635a15"},"timeout": 10,"pkt_type": "SetRPower","pkt_args": {"relay_index": 0,"level": 65535}}}'

Get State:
curl -XPUT http://127.0.0.1:6100/v1/lifx/command -HContent-Type:application/json -d '{"command": "query", "args": {"matcher": {"serial": "d073d5635a15"},"timeout": 10,"pkt_type": "GetRPower","pkt_args": {"relay_index": 0}}}'

Cheers,
Matt
